src/assets/TokensController.ts

```diff
@@ -505,66 +653,123 @@ export class TokensController extends BaseController<
+   * @param params - Object that holds token params.
+   * @param params.newTokens - The new tokens to set for the current network and selected account.
+   * @param params.newIgnoredTokens - The new ignored tokens to set for the current network and selected account.
+   * @param params.newDetectedTokens - The new detected tokens to set for the current network and selected account.
    * @returns The updated `allTokens` and `allIgnoredTokens` state.
    */
-  _getNewAllTokensState(newTokens: Token[], newIgnoredTokens: string[]) {
-    const { allTokens, allIgnoredTokens } = this.state;
+  _getNewAllTokensState(params: {
+    newTokens?: Token[];
+    newIgnoredTokens?: string[];
+    newDetectedTokens?: Token[];
+  }) {
+    const { newTokens, newIgnoredTokens, newDetectedTokens } = params;
+    const { allTokens, allIgnoredTokens, allDetectedTokens } = this.state;
     const { chainId, selectedAddress } = this.config;
-    const networkTokens = allTokens[chainId];
-    const networkIgnoredTokens = allIgnoredTokens[chainId];
 
-    const newNetworkTokens = {
-      ...networkTokens,
-      ...{ [selectedAddress]: newTokens },
-    };
-    const newIgnoredNetworkTokens = {
-      ...networkIgnoredTokens,
-      ...{ [selectedAddress]: newIgnoredTokens },
-    };
+    let newAllTokens = allTokens;
+    if (newTokens) {
+      const networkTokens = allTokens[chainId];
+      const newNetworkTokens = {
+        ...networkTokens,
+        ...{ [selectedAddress]: newTokens },
+      };
+      newAllTokens = {
+        ...allTokens,
+        ...{ [chainId]: newNetworkTokens },
+      };
+    }
 
-    const newAllTokens = {
-      ...allTokens,
-      ...{ [chainId]: newNetworkTokens },
-    };
+    let newAllIgnoredTokens = allIgnoredTokens;
+    if (newIgnoredTokens) {
+      const networkIgnoredTokens = allIgnoredTokens[chainId];
+      const newIgnoredNetworkTokens = {
+        ...networkIgnoredTokens,
+        ...{ [selectedAddress]: newIgnoredTokens },
+      };
+      newAllIgnoredTokens = {
+        ...allIgnoredTokens,
+        ...{ [chainId]: newIgnoredNetworkTokens },
+      };
+    }
 
-    const newAllIgnoredTokens = {
-      ...allIgnoredTokens,
-      ...{ [chainId]: newIgnoredNetworkTokens },
-    };
-    return { newAllTokens, newAllIgnoredTokens };
+    let newAllDetectedTokens = allDetectedTokens;
+    if (newDetectedTokens) {
+      const networkDetectedTokens = allDetectedTokens[chainId];
+      const newDetectedNetworkTokens = {
+        ...networkDetectedTokens,
+        ...{ [selectedAddress]: newDetectedTokens },
+      };
+      newAllDetectedTokens = {
+        ...allDetectedTokens,
+        ...{ [chainId]: newDetectedNetworkTokens },
+      };
+    }
+    return { newAllTokens, newAllIgnoredTokens, newAllDetectedTokens };
+  }
+
+  /**
+   * Fetch metadata for a token.
+   *
+   * @param tokenAddress - The address of the token.
+   * @returns The token metadata.
+   */
+  async fetchTokenMetadata(
```

Comment:
Should this method be public? :thinking: It seems like it was only meant for internal use, since the error handling was fairly specific to how it's being used here.

For example, lets say you wanted to fetch token metadata temporarily and show it in the UI without adding it to the user's list. You might want a method like this to do that. But if the user changes networks while the token metadata is being fetched, returning `null` here isn't very useful. Instead you'd want to retry. But returning `null` does make sense in the context of `addToken`, because upon a network switch we would want the whole "add token" operation would fail.

So maybe it makes more sense for this to be private. If we want to fetch token metadata elsewhere, they can call the `fetchTokenMetadata` function on the token-service module directly.

Reply:
Agreed. In addition to making this method private,  I've also improved the error handling logic to only return null while on non-supported networks. Otherwise, it will throw on any other error. Changes can be seen here - https://github.com/MetaMask/controllers/pull/808/commits/518afffe4a15bbae83d022c10ccb80b5bff06f2b